source/module_esolver/esolver_ks_lcao_elec.cpp

```diff
@@ -15,597 +15,589 @@
+                                                         GlobalC::ORB.get_rcutmax_Phi(),
+                                                         GlobalC::ucell.infoNL.get_rcutmax_Beta(),
+                                                         GlobalV::GAMMA_ONLY_LOCAL);
+
+        atom_arrange::search(GlobalV::SEARCH_PBC,
+                             GlobalV::ofs_running,
+                             GlobalC::GridD,
+                             GlobalC::ucell,
+                             GlobalV::SEARCH_RADIUS,
+                             GlobalV::test_atom_input,
+                             1);
         return;
     }
-    template <>
-    void ESolver_KS_LCAO<double, double>::get_S()
+
+    this->beforesolver(istep);
+    // pelec should be initialized before these calculations
+    this->pelec->init_scf(istep, this->sf.strucFac);
+    // self consistent calculations for electronic ground state
+    if (GlobalV::CALCULATION == "nscf")
     {
-        ModuleBase::TITLE("ESolver_KS_LCAO", "get_S");
-        ModuleBase::WARNING_QUIT("ESolver_KS_LCAO<TK, TR>::get_S", "not implemented for");
+        this->nscf();
     }
-    template <>
-    void ESolver_KS_LCAO<std::complex<double>, double>::get_S()
+    else if (GlobalV::CALCULATION == "get_pchg")
     {
-        ModuleBase::TITLE("ESolver_KS_LCAO", "get_S");
-            // (1) Find adjacent atoms for each atom.
-            GlobalV::SEARCH_RADIUS = atom_arrange::set_sr_NL(
-                GlobalV::ofs_running,
-                GlobalV::OUT_LEVEL,
-                GlobalC::ORB.get_rcutmax_Phi(),
-                GlobalC::ucell.infoNL.get_rcutmax_Beta(),
-                GlobalV::GAMMA_ONLY_LOCAL);
-
-            atom_arrange::search(
-                GlobalV::SEARCH_PBC,
-                GlobalV::ofs_running,
-                GlobalC::GridD,
-                GlobalC::ucell,
-                GlobalV::SEARCH_RADIUS,
-                GlobalV::test_atom_input);
-
-            this->RA.for_2d(this->orb_con.ParaV, GlobalV::GAMMA_ONLY_LOCAL);
-            this->UHM.genH.LM->ParaV = &this->orb_con.ParaV;
-            if(this->p_hamilt == nullptr)
-            {
-                this->p_hamilt = new hamilt::HamiltLCAO<std::complex<double>, double>(this->UHM.genH.LM, this->kv);
-                dynamic_cast<hamilt::OperatorLCAO<std::complex<double>, double>*>(this->p_hamilt->ops)->contributeHR();
-            }
-            ModuleIO::output_S_R(this->UHM, this->p_hamilt, "SR.csr");
+        IState_Charge ISC(this->psi, this->LOC);
+        ISC.begin(this->UHM.GG,
+                  this->pelec,
+                  this->pw_rho,
+                  this->pw_big,
+                  GlobalV::GAMMA_ONLY_LOCAL,
+                  GlobalV::NBANDS_ISTATE,
+                  GlobalV::NBANDS,
+                  GlobalV::nelec,
+                  GlobalV::NSPIN,
+                  GlobalV::global_out_dir,
+                  GlobalV::MY_RANK,
+                  GlobalV::ofs_warning);
```

Comment:
These are the only lines changed in the file esolver_ks_lcao_elec.cpp, which are the input parameters of the begin function. According to the modifications made to the function definition, corresponding modifications have also been made where the function is used. Apologize for any inconvenience caused to the reviewers🥺